Syscall handler cleanups

tyrantlizard77 opened the pull request:
The macros in syscall_handler.h have been updated and extended for clarity.

A C API only used in handlers moved to the proper header.

Syscall handlers updated to use the new macros.

Commit:
syscalls: add _SYSCALL_VERIFY_MSG()

Expecting stringified expressions to be completely comprehensible to end
users is wishful thinking; we really need to express what a failed
system call verification step means in human terms in most cases.

Memory buffer and kernel object checks now are implemented in terms of
_SYSCALL_VERIFY_MSG.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
userspace: move _k_object_validate() definition

This API only gets used inside system call handlers and a specific test
case dedicated to it. Move definition to the private kernel header along
with the rest of the defines for system call handlers.

A non-userspace inline variant of this function is unnecessary and has
been deleted.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
syscall_handler: introduce new macros

Instead of boolean arguments to indicate memory read/write
permissions, or init/non-init APIs, new macros are introduced
which bake the semantics directly into the name of the macro.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: system call handler cleanup

Use new _SYSCALL_OBJ/_SYSCALL_OBJ_INIT macros.

Use new _SYSCALL_MEMORY_READ/_SYSCALL_MEMORY_WRITE macros.

Some non-obvious checks changed to use _SYSCALL_VERIFY_MSG.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

lpereira commented on kernel/msg_q.c:
```diff
@@ -69,8 +69,8 @@ u32_t _handler_k_msgq_init(u32_t q, u32_t buffer, u32_t msg_size,
-	_SYSCALL_IS_OBJ(q, K_OBJ_MSGQ, 1, ssf);
-	_SYSCALL_MEMORY(buffer, msg_size * max_msgs, 1, ssf);
+	_SYSCALL_OBJ_INIT(q, K_OBJ_MSGQ, ssf);
+	_SYSCALL_MEMORY_WRITE(buffer, msg_size * max_msgs, ssf);
```
Unrelated to the patch, but if `msg_size * max_msgs` overflows, since they're both unsigned, this check can be bypassed.  Better to use `__builtin_mul_overflow()`, maybe wrapped in a `_SYSCALL_MEMORY_WRITE_ARRAY()` macro.

lpereira replied:
There are other instances where a multiplication is performed, which could lead to the same problem.

andrewboie replied:
i'll add this to this PR good catch

andrewboie commented:
Note to self: CI error is need to fix include path in obj_validation test case

```
 /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/tests/kernel/mem_protect/obj_validation/src/main.c: In function ‘test_bad_object’:
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/tests/kernel/mem_protect/obj_validation/src/main.c:26:10: error: implicit declaration of function ‘_k_object_validate’ [-Werror=implicit-function-declaration]
  return !_k_object_validate(sem, K_OBJ_SEM, 0);
          ^~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
```

Commit:
syscall_handler: handle multiplication overflow

Computing the total size of the array need to handle the case where
the product overflow a 32-bit unsigned integer.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
fixed array bounds multiplication issue with an additional patch

lpereira commented on kernel/include/syscall_handler.h:
```diff
@@ -116,6 +116,47 @@ int _k_object_validate(void *obj, enum k_objects otype, int init);
+#define _SYSCALL_MEMORY_ARRAY(ptr, nmemb, size, write, ssf) \
+	do { \
+		u32_t product; \
+		_SYSCALL_VERIFY_MSG(!__builtin_umul_overflow((u32_t)nmemb, \
```
Put parenthesis around `nmemb` and `size`, as they're likely to be used in expressions, and you want the whole expression cast as `u32_t`.

andrewboie commented:
Added parens around macro arguments.
CI failure fixed by #4275

lpereira reviewed (approved):
This is better, thanks!

andrewboie commented:
CI issue should be fixed by #4280 I'll rebase once that goes in.
